### PR TITLE
PUBDEV-8977: add accessor functions in R and Python to access the GLM neg logll and average objective

### PR DESCRIPTION
**Pull request by wendycwong**
This PR complete the task in JIRA: https://h2oai.atlassian.net/browse/PUBDEV-8977

I added two accessor functions in R and Python to make sure GLM objective (average) and negative log likelihood are accessible to them when generate_scoring_history = True/TRUE:

negative_log_likelihood
average_objective

This is ugly I agree because I have to get it from scoring_history.  I am putting a JIRA in to add negative_log_likelhood and average_objective to be part of the modelMetrics.  Once that is done, it will get better.

### CODE REVIEW

File: h2o-py/h2o/model/model_base.py [hunk 1]
```diff
@@ -539,6 +539,46 @@ def scoring_history(self):
...
+                        scLen = len(sc_history._cell_values)
+                        return sc_history._cell_values[scLen-1][index_val]
+                    else:
+                        print("{0} not available.".format(value))
+                else:
+                    raise H2OValueError("GLM scoring_history is missing.  Cannot extract {0}.".value)
+            else:
+                raise H2OValueError("For now, need to set generate_scoring_history to True to get {0}".format(value))
+        else:
+            raise H2OValueError("{0} is only for GLM models.".format(value))
```
Comment (tomasfryda) [inline h2o-py/h2o/model/model_base.py:579]: I don't understand why this should be limited just for glm. The scoring history is `H2OTwoDimTable` for all the models so this method could be useful for other models as well.

If you want to keep it just for GLM, I'd recommend to implement it as an extension - create a mixin class in `h2o-py/h2o/model/extensions` and then register it in `h2o-bindings/bin/custom/python/gen_glm.py` https://github.com/h2oai/h2o-3/blob/a81177da044cae1cc340fbcba469c11eaff9b9e9/h2o-bindings/bin/custom/python/gen_glm.py#L1-L7

Comment (wendycwong) [reply h2o-py/h2o/model/model_base.py:579]: I agree, I don't want to limit this function to just GLM.  However for now I am doing this as a quick fix.  I have opened an JIRA to resolve this issue at least for GLM but I think you are right, this needs to be added to more models.

File: h2o-py/h2o/model/model_base.py [hunk 2]
```diff
@@ -539,6 +539,46 @@ def scoring_history(self):
...
+        
+    def negative_log_likelihood(self):
+        """
+        Retrieve model negative likelihood function value from scoring history if exists for GLM model
+        
+        :return: the negative likelihood function value
+        """
+        return self.extract_scoring_history("negative_log_likelihood")
+
+    def avarege_objective(self):
```
[the remark this answers is not in the document]
Comment (wendycwong) [reply h2o-py/h2o/model/model_base.py]: I caught this one too.

File: h2o-py/h2o/model/model_base.py (hunk 1 above)
Comment (wendycwong) [reply h2o-py/h2o/model/model_base.py:579]: The loglikelihood and objective functions should have been part of the model metrics and not for us to get it from scoring_history at all.  The two values are evaluated during the scoring part anyway, we can just add it to the list of model metrics. In the long run, I do not want to get these two values out of the scoring history at all.

Comment (tomasfryda) [reply h2o-py/h2o/model/model_base.py:579]: Then for the quick fix, I'd suggest to make this method "private" (prefix with underscore) for now. (Now it's added to `model_base` so user of any model could be tempted to use it.)
In R, you don't export this function as well.

File: h2o-r/h2o-package/R/models.R [hunk 3]
```diff
@@ -2498,6 +2498,71 @@ h2o.get_regression_influence_diagnostics <- function(model, predictorSize = -1)
...
+#' @export 
+h2o.negative_log_likelihood <- function(model) {
+    return(h2o.extract_value(model, "negative_log_likelihood"))
+}
+
+#'
+#' Extracts the final training average objective function of a GLM model.
+#'
+#' @param model an \linkS4class{H2OModel} object.
+#'
```
Comment (tomasfryda) [inline h2o-r/h2o-package/R/models.R:2531]: CRAN might be require return value specification as well.

File: h2o-r/h2o-package/R/models.R [hunk 4]
```diff
@@ -2498,6 +2498,71 @@ h2o.get_regression_influence_diagnostics <- function(model, predictorSize = -1)
     }
 }
 
+#'
+#' Extracts the final training negative log likelihood of a GLM model.
+#'
+#' @param model an \linkS4class{H2OModel} object.
+#'
```
Comment (tomasfryda) [inline h2o-r/h2o-package/R/models.R:2505]:
```suggestion
#' @return The final training negative log likelihood of a GLM model.
```

File: h2o-r/h2o-package/R/models.R (hunk 3 above)
Comment (wendycwong) [reply h2o-r/h2o-package/R/models.R:2533]: Added it.  Thanks for the catch

File: h2o-py/h2o/model/model_base.py (hunk 1 above)
Comment (wendycwong) [reply h2o-py/h2o/model/model_base.py]: true.  Done.

**Review by tomasfryda — APPROVED** [review]
Looks good to me, thank you @wendycwong !

File: h2o-py/tests/testdir_algos/glm/pyunit_PUBDEV_8977_negLL_Obj.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    glm_model.train(x=my_x, y=my_y, training_frame=d)
+    glm_model_noReg = H2OGeneralizedLinearEstimator(family="gaussian", seed=1234, generate_scoring_history=True, lambda_=0.0)
+    glm_model_noReg.train(x=my_x, y=my_y, training_frame=d)
+    nll = glm_model.negative_log_likelihood()
+    obj = glm_model.average_objective()
+    nll_noReg = glm_model_noReg.negative_log_likelihood()
+    obj_noReg = glm_model_noReg.average_objective()
+    assert abs(nll_noReg-obj_noReg/glm_model_noReg.actual_params["obj_reg"]) < 1e-6, \
+        "objective ({0}) and negative log likelihood ({1}) should equal but do not.".format(obj_noReg, nll_noReg)
+    assert abs(nll-obj) > 1e-6, "objective ({0}) and negative log likelihood ({1}) should not equal but do" \
```
Comment (mn-mikke) [inline h2o-py/tests/testdir_algos/glm/pyunit_PUBDEV_8977_negLL_Obj.py:24]: I think there is a typo in the error message (opposite meaning):
`but do not` -> `they do`

Comment (wendycwong) [reply h2o-py/tests/testdir_algos/glm/pyunit_PUBDEV_8977_negLL_Obj.py:24]: The error message will only show up when the test failed meaning the two values should equal but in reality they do not.  So the message is correct.  I know it is confusing.  Sorry.

Comment (mn-mikke) [reply h2o-py/tests/testdir_algos/glm/pyunit_PUBDEV_8977_negLL_Obj.py:24]: > the two values should equal but in reality they do not.

IMHO, this holds only for assert on line 22. On line 24 you're asserting that the values are different. And the current message is:

objective ({0}) and negative log likelihood ({1}) should not equal but _**do not.**_

Anyway, it's just a minor thing.

**Pull request merged**